src/Analyzers/Core/Analyzers/AbstractBuiltInUnnecessaryCodeStyleDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+
+#if CODE_STYLE
+using OptionSet = Microsoft.CodeAnalysis.Diagnostics.AnalyzerConfigOptions;
+#endif
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    /// <summary>
+    /// Code style analyzer that reports at least one 'unnecessary' code diagnostic.
+    /// </summary>
+    internal abstract class AbstractBuiltInUnnecessaryCodeStyleDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        /// <summary>
+        /// Constructor for an unnecessary code style analyzer with a single diagnostic descriptor and
+        /// unique <see cref="IPerLanguageOption"/> code style option.
+        /// </summary>
+        /// <param name="diagnosticId">Diagnostic ID reported by this analyzer</param>
+        /// <param name="enforceOnBuild">Build enforcement recommendation for this analyzer</param>
+        /// <param name="option">
+        /// Per-language option that can be used to configure the given <paramref name="diagnosticId"/>.
+        /// <see langword="null"/>, if there is no such unique option.
+        /// </param>
+        /// <param name="fadingOption">
+        /// Per-language fading option that can be used to configure if the diagnostic should be faded in the IDE or not.
+        /// <see langword="null"/>, if there is no such unique fading option.
+        /// </param>
+        /// <param name="title">Title for the diagnostic descriptor</param>
+        /// <param name="messageFormat">
+        /// Message for the diagnostic descriptor.
+        /// <see langword="null"/> if the message is identical to the title.
+        /// </param>
+        /// <param name="configurable">Flag indicating if the reported diagnostics are configurable by the end users</param>
+        protected AbstractBuiltInUnnecessaryCodeStyleDiagnosticAnalyzer(
```

Comment:
>AbstractBuiltInUnnecessaryCodeStyleDiagnosticAnalyzer [](http://example.com/codeflow?start=18&length=53)

Can we merge all these overloads (and the ones on the base class as well) to avoid so much repetition?

Reply:
i am genuinely surprised we need so many overloads (in both types).